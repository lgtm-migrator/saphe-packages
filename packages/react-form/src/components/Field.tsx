import React, { ChangeEvent, FocusEvent, ReactElement } from 'react';
import {
  FieldTypes,
  ICheckBoxField,
  IField,
  INumberField,
  ISelectField,
  ITextAreaField,
  ITextField,
} from '../utils/fieldTypes';
import { AddFieldPack, HTMLField } from '../utils/helperTypes';
import CheckBoxField from './fields/CheckBoxField';
import NumberField from './fields/NumberField';
import SelectField from './fields/SelectField';
import TextAreaField from './fields/TextAreaField';
import TextField from './fields/TextField';

interface Props<T extends IField> {
  id: string;
  field: T;
  name: string;
  error: string;
  value: string;
  onChange: (e: ChangeEvent<HTMLField>) => void;
  onBlur: (e: FocusEvent<HTMLField>) => void;
}

export default function Field<T extends IField>(
  props: AddFieldPack<Props<T>>,
): ReactElement {
  switch (props.field.type) {
    case FieldTypes.TEXT:
      return <TextField {...props} {...(props.field as ITextField)} />;
    case FieldTypes.TEXTAREA:
      return <TextAreaField {...props} {...(props.field as ITextAreaField)} />;
    case FieldTypes.SELECT:
      return <SelectField {...props} {...(props.field as ISelectField)} />;
    case FieldTypes.CHECKBOX:
      return <CheckBoxField {...props} {...(props.field as ICheckBoxField)} />;
    case FieldTypes.NUMBER:
      return <NumberField {...props} {...(props.field as INumberField)} />;
  }
}
