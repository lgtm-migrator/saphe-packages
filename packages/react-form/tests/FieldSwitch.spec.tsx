import React from 'react';
import Field from '../src/components/FieldSwitch';
import {
  matchSnapshot,
  minimalCheckBoxField,
  minimalNumberField,
  minimalSelectField,
  minimalTextAreaField,
  minimalTextField,
} from './testHelpers';

const dummyProps = {
  id: 'formNameFieldName',
  name: 'fieldName',
  label: 'Field Label',
  description: '',
  describedBy: 'formNameFieldNameDescription',
  disabled: false,
  error: '',
  value: '',
  onChange: jest.fn(),
  onBlur: jest.fn(),
};

describe('Field', () => {
  it('renders a text field correctly', () => {
    matchSnapshot(<Field field={minimalTextField} {...dummyProps} />);
    matchSnapshot(
      <Field
        field={{ ...minimalTextField, description: 'description' }}
        {...dummyProps}
        error="error"
      />,
    );
  });

  it('renders a text area field correctly', () => {
    matchSnapshot(<Field field={minimalTextAreaField} {...dummyProps} />);
    matchSnapshot(
      <Field
        field={{
          ...minimalTextAreaField,
          description: 'description',
          rows: 12,
        }}
        {...dummyProps}
        error="error"
      />,
    );
  });

  it('renders a select field correctly', () => {
    matchSnapshot(<Field field={minimalSelectField} {...dummyProps} />);
    matchSnapshot(
      <Field
        field={{
          ...minimalSelectField,
          description: 'description',
          placeholder: 'placeholder',
        }}
        {...dummyProps}
        error="error"
      />,
    );
  });

  it('renders a checkbox field correctly', () => {
    matchSnapshot(
      <Field field={minimalCheckBoxField} {...dummyProps} value={false} />,
    );
    matchSnapshot(
      <Field
        field={{ ...minimalCheckBoxField, description: 'description' }}
        {...dummyProps}
        value={false}
        error="error"
      />,
    );
  });

  it('renders a number field correctly', () => {
    matchSnapshot(
      <Field field={minimalNumberField} {...dummyProps} value={10} />,
    );
    matchSnapshot(
      <Field
        field={{ ...minimalNumberField, description: 'description' }}
        {...dummyProps}
        value={10}
        error="error"
      />,
    );
  });
});
