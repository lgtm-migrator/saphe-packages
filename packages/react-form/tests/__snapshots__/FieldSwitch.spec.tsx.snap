// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field renders a check field correctly 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": "25px",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
      }
    }
  >
    <input
      aria-describedby="formNameFieldNameDescription"
      disabled={false}
      id="formNameFieldName"
      name="fieldName"
      onBlur={[MockFunction]}
      onChange={[Function]}
      style={
        Object {
          "borderColor": "#767676",
          "borderRadius": "3px",
          "borderStyle": "solid",
          "borderWidth": "1px",
          "marginBottom": "3px",
          "marginRight": "10px",
        }
      }
      type="checkbox"
      value="false"
    />
    <label
      htmlFor="formNameFieldName"
    >
      Check Field
    </label>
  </div>
  
  
</div>
`;

exports[`Field renders a check field correctly 2`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": "25px",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
      }
    }
  >
    <input
      aria-describedby="formNameFieldNameDescription"
      disabled={false}
      id="formNameFieldName"
      name="fieldName"
      onBlur={[MockFunction]}
      onChange={[Function]}
      style={
        Object {
          "borderColor": "red",
          "borderRadius": "3px",
          "borderStyle": "solid",
          "borderWidth": "1px",
          "marginBottom": "3px",
          "marginRight": "10px",
        }
      }
      type="checkbox"
      value="false"
    />
    <label
      htmlFor="formNameFieldName"
    >
      Check Field
    </label>
  </div>
  <p
    data-testid="formNameFieldNameError"
    style={
      Object {
        "color": "red",
        "margin": 0,
      }
    }
  >
    error
  </p>
  <p
    id="formNameFieldNameDescription"
    style={
      Object {
        "margin": 0,
      }
    }
  >
    description
  </p>
</div>
`;

exports[`Field renders a number field correctly 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": "25px",
    }
  }
>
  <label
    htmlFor="formNameFieldName"
  >
    Number Field
  </label>
  <input
    aria-describedby="formNameFieldNameDescription"
    disabled={false}
    id="formNameFieldName"
    name="fieldName"
    onBlur={[MockFunction]}
    onChange={[Function]}
    style={
      Object {
        "borderColor": "#767676",
        "borderRadius": "3px",
        "borderStyle": "solid",
        "borderWidth": "1px",
        "marginBottom": "3px",
      }
    }
    type="number"
    value={10}
  />
  
  
</div>
`;

exports[`Field renders a number field correctly 2`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": "25px",
    }
  }
>
  <label
    htmlFor="formNameFieldName"
  >
    Number Field
  </label>
  <input
    aria-describedby="formNameFieldNameDescription"
    disabled={false}
    id="formNameFieldName"
    name="fieldName"
    onBlur={[MockFunction]}
    onChange={[Function]}
    style={
      Object {
        "borderColor": "red",
        "borderRadius": "3px",
        "borderStyle": "solid",
        "borderWidth": "1px",
        "marginBottom": "3px",
      }
    }
    type="number"
    value={10}
  />
  <p
    data-testid="formNameFieldNameError"
    style={
      Object {
        "color": "red",
        "margin": 0,
      }
    }
  >
    error
  </p>
  <p
    id="formNameFieldNameDescription"
    style={
      Object {
        "margin": 0,
      }
    }
  >
    description
  </p>
</div>
`;

exports[`Field renders a select field correctly 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": "25px",
    }
  }
>
  <label
    htmlFor="formNameFieldName"
  >
    Select Field
  </label>
  <select
    aria-describedby="formNameFieldNameDescription"
    disabled={false}
    id="formNameFieldName"
    name="fieldName"
    onBlur={[MockFunction]}
    onChange={[Function]}
    style={
      Object {
        "borderColor": "#767676",
        "borderRadius": "3px",
        "borderStyle": "solid",
        "borderWidth": "1px",
        "marginBottom": "3px",
      }
    }
    value=""
  >
    <option
      disabled={true}
      value=""
    >
      
    </option>
    <option
      value="option1"
    >
      Option 1
    </option>
    <option
      value="option2"
    >
      Option 2
    </option>
  </select>
  
  
</div>
`;

exports[`Field renders a select field correctly 2`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": "25px",
    }
  }
>
  <label
    htmlFor="formNameFieldName"
  >
    Select Field
  </label>
  <select
    aria-describedby="formNameFieldNameDescription"
    disabled={false}
    id="formNameFieldName"
    name="fieldName"
    onBlur={[MockFunction]}
    onChange={[Function]}
    style={
      Object {
        "borderColor": "red",
        "borderRadius": "3px",
        "borderStyle": "solid",
        "borderWidth": "1px",
        "marginBottom": "3px",
      }
    }
    value=""
  >
    <option
      disabled={true}
      value=""
    >
      placeholder
    </option>
    <option
      value="option1"
    >
      Option 1
    </option>
    <option
      value="option2"
    >
      Option 2
    </option>
  </select>
  <p
    data-testid="formNameFieldNameError"
    style={
      Object {
        "color": "red",
        "margin": 0,
      }
    }
  >
    error
  </p>
  <p
    id="formNameFieldNameDescription"
    style={
      Object {
        "margin": 0,
      }
    }
  >
    description
  </p>
</div>
`;

exports[`Field renders a text area field correctly 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": "25px",
    }
  }
>
  <label
    htmlFor="formNameFieldName"
  >
    Text Area Field
  </label>
  <textarea
    aria-describedby="formNameFieldNameDescription"
    disabled={false}
    id="formNameFieldName"
    name="fieldName"
    onBlur={[MockFunction]}
    onChange={[Function]}
    rows={6}
    style={
      Object {
        "borderColor": "#767676",
        "borderRadius": "3px",
        "borderStyle": "solid",
        "borderWidth": "1px",
        "marginBottom": "3px",
      }
    }
    value=""
  />
  
  
</div>
`;

exports[`Field renders a text area field correctly 2`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": "25px",
    }
  }
>
  <label
    htmlFor="formNameFieldName"
  >
    Text Area Field
  </label>
  <textarea
    aria-describedby="formNameFieldNameDescription"
    disabled={false}
    id="formNameFieldName"
    name="fieldName"
    onBlur={[MockFunction]}
    onChange={[Function]}
    rows={12}
    style={
      Object {
        "borderColor": "red",
        "borderRadius": "3px",
        "borderStyle": "solid",
        "borderWidth": "1px",
        "marginBottom": "3px",
      }
    }
    value=""
  />
  <p
    data-testid="formNameFieldNameError"
    style={
      Object {
        "color": "red",
        "margin": 0,
      }
    }
  >
    error
  </p>
  <p
    id="formNameFieldNameDescription"
    style={
      Object {
        "margin": 0,
      }
    }
  >
    description
  </p>
</div>
`;

exports[`Field renders a text field correctly 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": "25px",
    }
  }
>
  <label
    htmlFor="formNameFieldName"
  >
    Text Field
  </label>
  <input
    aria-describedby="formNameFieldNameDescription"
    autoComplete="off"
    disabled={false}
    id="formNameFieldName"
    name="fieldName"
    onBlur={[MockFunction]}
    onChange={[Function]}
    style={
      Object {
        "borderColor": "#767676",
        "borderRadius": "3px",
        "borderStyle": "solid",
        "borderWidth": "1px",
        "marginBottom": "3px",
      }
    }
    type="text"
    value=""
  />
  
  
</div>
`;

exports[`Field renders a text field correctly 2`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "marginBottom": "25px",
    }
  }
>
  <label
    htmlFor="formNameFieldName"
  >
    Text Field
  </label>
  <input
    aria-describedby="formNameFieldNameDescription"
    autoComplete="off"
    disabled={false}
    id="formNameFieldName"
    name="fieldName"
    onBlur={[MockFunction]}
    onChange={[Function]}
    style={
      Object {
        "borderColor": "red",
        "borderRadius": "3px",
        "borderStyle": "solid",
        "borderWidth": "1px",
        "marginBottom": "3px",
      }
    }
    type="text"
    value=""
  />
  <p
    data-testid="formNameFieldNameError"
    style={
      Object {
        "color": "red",
        "margin": 0,
      }
    }
  >
    error
  </p>
  <p
    id="formNameFieldNameDescription"
    style={
      Object {
        "margin": 0,
      }
    }
  >
    description
  </p>
</div>
`;
